listenbrainz/webserver/static/js/src/recommendations/RecommendationCard.test.tsx

```diff
@@ -16,25 +22,297 @@ const recommendation: Recommendation = {
+    await instance.submitFeedback("hate");
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith("lalala", "yyyy", "hate");
+
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(1);
+    expect(instance.props.updateFeedback).toHaveBeenCalledWith("yyyy", "hate");
+
+    // Emulate updating the props after calling Recommendations component's updateFeedback
+    wrapper.setProps({ currentFeedback: "hate" });
+
+    // Check that HTML elements have changed accordingly
+    expect(
+      wrapper.find(".recommendation-controls").childAt(0).hasClass("hate")
+    ).toEqual(true);
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(0).prop("icon").iconName).toEqual("angry");
+    expect(wrapper.find('.recommendation-controls').childAt(0).childAt(2).text()).toEqual("Hate");
+  });
+
+  it("does nothing if isCurrentUser is false", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard
+        {...{ ...props, isCurrentUser: false, updateFeedback: jest.fn() }}
+      />
+    );
+    const instance = wrapper.instance();
+
+    const spy = jest.spyOn(instance.APIService, "submitRecommendationFeedback");
+    spy.mockImplementation(() => Promise.resolve(200));
+
+    instance.submitFeedback("dislike");
+    expect(spy).toHaveBeenCalledTimes(0);
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(0);
+    expect(wrapper.exists(".recommendation-controls")).toEqual(false);
+  });
+
+  it("does nothing if CurrentUser.authtoken is not set", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard
+        {...{
+          ...props,
+          currentUser: { auth_token: undefined, name: "test" },
+          updateFeedback: jest.fn(),
+        }}
+      />
+    );
+    const instance = wrapper.instance();
+
+    const spy = jest.spyOn(instance.APIService, "submitRecommendationFeedback");
+    spy.mockImplementation(() => Promise.resolve(200));
+
+    instance.submitFeedback("love");
+    expect(spy).toHaveBeenCalledTimes(0);
+    expect(instance.props.updateFeedback).toHaveBeenCalledTimes(0);
+    expect(wrapper.exists(".recommendation-controls")).toEqual(false);
+  });
+
+  it("doesn't call updateFeedback if status code is not 200", async () => {
+    const wrapper = shallow<RecommendationCard>(
+      <RecommendationCard {...props} />
+    );
+    const instance = wrapper.instance();
+    props.updateFeedback = jest.fn();
+
+    const spy = jest.spyOn(instance.APIService, "submitRecommendationFeedback");
+    spy.mockImplementation(() => Promise.resolve(201));
+
+    instance.submitFeedback("bad_recommendation");
+
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(spy).toHaveBeenCalledWith("lalala", "yyyy", "bad_recommendation");
+
+    expect(props.updateFeedback).toHaveBeenCalledTimes(0);
+    expect(
+      wrapper.find(".recommendation-controls").childAt(0).hasClass("love")
```

Comment:
I think we don't need to check HTML elements at places where updateFeedback is not called at all. What do you think?

Reply:
I guess it's a good way to ensure that the feedback hasn't changed, but perhaps it is better done by checking the value of `instance.props.currentFeedback` instead, especially if you are making separate tests for the soul of the PR, as you said :)